ui/app/components/auth/form/oidc-jwt.ts

```diff
@@ -0,0 +1,280 @@
+      });
+      // responsible for redirect after auth data is persisted
+      this.handleAuthResponse(authResponse);
+    } catch (error) {
+      this.onError(error as Error);
+    }
+  }
+
+  // * OIDC AUTH PART 1
+  // 1. request oidc/auth_url to check for config errors, if none continue
+  // 2. open popup window at auth_url
+  async startOIDCAuth() {
+    await this.fetchRole.perform();
+
+    const error =
+      this.fetchedRole && !this.fetchedRole.authUrl
+        ? 'Missing auth_url. Please check that allowed_redirect_uris for the role include this mount path.'
+        : this.errorMessage || null;
+
+    if (error) {
+      this.onError(error);
+    } else {
+      const win = window;
+      const POPUP_WIDTH = 500;
+      const POPUP_HEIGHT = 600;
+      const left = win.screen.width / 2 - POPUP_WIDTH / 2;
+      const top = win.screen.height / 2 - POPUP_HEIGHT / 2;
+      const oidcWindow = win.open(
+        this.fetchedRole?.authUrl,
+        'vaultOIDCWindow',
+        `width=${POPUP_WIDTH},height=${POPUP_HEIGHT},resizable,scrollbars=yes,top=${top},left=${left}`
+      );
+
+      this.prepareForOIDC.perform(oidcWindow);
+    }
+  }
+
+  // * OIDC AUTH PART 2
+  // 3. watch popups for premature closure
+  // 4. wait message event from window.postMessage() in oidc-callback route
+  prepareForOIDC = task(async (oidcWindow) => {
+    // NOTE TO DEVS: Be careful when updating the OIDC flow and ensure the updates
+    // work with implicit flow. See issue https://github.com/hashicorp/vault-plugin-auth-jwt/pull/192
+    const thisWindow = window;
+
+    // start watching the popup window and the current one
+    this.watchPopup.perform(oidcWindow);
+    this.watchCurrent.perform(oidcWindow);
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      // wait for message posted from oidc callback, see issue https://github.com/hashicorp/vault/issues/12436
+      // ensure that postMessage event is from expected source
+      const event = (await waitForEvent(thisWindow, 'message')) as unknown as MessageEvent;
+      if (event.origin === thisWindow.origin && event.isTrusted && event.data.source === 'oidc-callback') {
+        // event.data are params from the oidc callback url parsed by getParamsForCallback in the oidc-callback route
+        return this.exchangeOIDC.perform(event.data, oidcWindow);
+      }
+    }
+  });
+
+  // * OIDC AUTH PART 3
+  // 5. check parsed url for expected state params
+  // 6. if successful, request client_token from oidc/callback
+  // 7. close popups and continue login with client_token
+  exchangeOIDC = task(async (oidcState, oidcWindow) => {
+    if (oidcState === null || oidcState === undefined) {
+      return;
+    }
+
+    const { path, state, code } = oidcState;
+    if (!path || !state || !code) {
+      return this.cancelLogin(oidcWindow, ERROR_MISSING_PARAMS);
+    }
+
+    // TODO CMB - when wiring up components check if this is still necessary
```

Comment:
because query params get involved, waiting until this is fully wired up before testing and updating accordingly. (logging in with a namespace works with the current logic)